services/scheduler/scheduler.go

```diff
@@ -41,34 +43,47 @@ func processConfig(config *pb.Config, c pb.ContextBoxServiceClient) error {
 	return nil
 }
 
-//configProcessor is worker function which invokes the processConfig() function
-//returns func()error which will carry out the processConfig() execution
-func configProcessor(c pb.ContextBoxServiceClient) func() error {
-	return func() error {
-		//pull an item from a queue in cbox
-		res, err := cbox.GetConfigScheduler(c)
-		if err != nil {
-			return fmt.Errorf("error while getting Scheduler config: %w", err)
-		}
+// configProcessor will fetch new configs from the supplied connection
+// to the context-box service. Each received config will be processed in
+// a separate go-routine. If a sync.WaitGroup is supplied it will call
+// the Add(1) and then the Done() method on it after the go-routine finishes
+// the work, if nil it will be ignored.
+func configProcessor(c pb.ContextBoxServiceClient, wg *sync.WaitGroup) error {
+	//pull an item from a queue in cbox
+	res, err := cbox.GetConfigScheduler(c)
+	if err != nil {
+		return fmt.Errorf("error while getting Scheduler config: %w", err)
+	}
 
-		//process config
-		config := res.GetConfig()
-		if config != nil {
-			go func(config *pb.Config) {
-				log.Info().Msgf("Processing %s ", config.Name)
-				err := processConfig(config, c)
-				if err != nil {
-					log.Error().Msgf("processConfig() failed: %s", err)
-					//save error message to config
-					errSave := saveErrorMessage(config, c, err)
-					if errSave != nil {
-						log.Error().Msgf("scheduler:failed to save error to the config: %s : processConfig failed: %s", errSave, err)
-					}
-				}
-			}(config)
-		}
+	//process config
+	config := res.GetConfig()
+	if config == nil {
 		return nil
 	}
+
+	if wg != nil {
+		// we received a non-nil config thus we add a new worker to the wait group.
+		wg.Add(1)
+	}
+
+	go func() {
+		if wg != nil {
```

Comment:
is this second if necessary?

Reply:
Yes, calling wg.Done() on a nil receiver would panic

Reply:
Yes, but 3 lines above, you are already checking for a nill `wg`. Is there a chance that by the time it will trigger a goroutine, it will become nil?

Reply:
no, the WaitGroup will always outlive the spawned go-routines, I added this check if someone wanted to call the function without a WaitGroup, basically if the parameter is `nil` these calls will be ignored, if not `nil` they will use the WaitGroup